src/libsystemd/sd-device/test-sd-device.c

```diff
@@ -63,9 +64,81 @@ static void test_sd_device_basic(void) {
         }
 }
 
+static void test_sd_device_enumerator_filter_subsystem_one(const char *subsystem, Hashmap *h) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *d, *t;
+
+        assert_se(sd_device_enumerator_new(&e) >= 0);
+        assert_se(sd_device_enumerator_add_match_subsystem(e, subsystem, true) >= 0);
+
+        FOREACH_DEVICE(e, d) {
+                const char *syspath;
+
+                assert_se(sd_device_get_syspath(d, &syspath) >= 0);
+                assert_se(t = hashmap_remove(h, syspath));
+                assert_se(!sd_device_unref(t));
+
+                log_debug("Removed subsystem:%s syspath:%s", subsystem, syspath);
+        }
+
+        assert_se(hashmap_isempty(h));
+}
+
+static void test_sd_device_enumerator_filter_subsystem(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        Hashmap *subsystems, *h;
+        sd_device *d;
+        Iterator i;
+        char *s;
+
+        assert_se(subsystems = hashmap_new(&string_hash_ops));
+        assert_se(sd_device_enumerator_new(&e) >= 0);
+
+        FOREACH_DEVICE(e, d) {
+                const char *syspath, *subsystem;
+                int r;
+
+                assert_se(sd_device_get_syspath(d, &syspath) >= 0);
+
+                r = sd_device_get_subsystem(d, &subsystem);
+                assert_se(r >= 0 || r == -ENOENT);
+                if (r < 0)
+                        continue;
+
+                h = hashmap_get(subsystems, subsystem);
+                if (!h) {
+                        char *str;
+                        assert_se(str = strdup(subsystem));
+                        assert_se(h = hashmap_new(&string_hash_ops));
+                        assert_se(hashmap_put(subsystems, str, h) >= 0);
+                }
+
+                r = hashmap_put(h, syspath, d);
+                assert_se(r >= 0 || r == -EEXIST);
+                if (r < 0) {
+                        log_info("Duplicated subsystem:%s syspath:%s", subsystem, syspath);
+                        continue;
+                }
+
+                assert_se(sd_device_ref(d));
+
+                log_debug("Added subsystem:%s syspath:%s", subsystem, syspath);
+        }
+
+        HASHMAP_FOREACH_KEY(h, s, subsystems, i) {
+                test_sd_device_enumerator_filter_subsystem_one(s, h);
+                hashmap_free(h);
+                free(s);
+        }
```

Comment:
hmm, so you are freeing the elements of the hashtable as you iterate through it, but that means that after this loop passed all entries will still be there, but point to invalidated memory, which is super ugly I think.

I think this would be much nicer if this was a loop around hashmap_steal_first() or so, i.e. that we keep removing the first entry and free it, as we go. That way the hashtable is guaranteed to be empty at the end and not contain entries pointing into bad memory

Reply:
I may misunderstand your point. The reason I use `HASHMAP_FOREACH_KEY()` is that we do not have, say, `hashmap_steal_first_key_and_val()`.

> that means that after this loop passed all entries will still be there

Sorry, I cannot understand this... I tested this with valgrind, and the test has no memleak.

Reply:
Thanks. Now I catch your point. I will rework this.